File: heterogeneousadapter/src/main/java/com/iheartradio/heterogeneousadapter/dataset/ListDataSet.java
package com.iheartradio.heterogeneousadapter.dataset;


import java.util.ArrayList;
import java.util.List;

public final class ListDataSet<T> implements DataSet<T> {

    private List<T> mData;


    public static <T> ListDataSet<T> of(final List<T> other) {
        return new ListDataSet<>(other);
    }

    public ListDataSet() {
        this(new ArrayList<T>());
    }

    public ListDataSet(final List<? extends T> data) {
        setData(data);
    }

    public void setData(final List<? extends T> newData) {
        mData = new ArrayList<>(newData);
    }

    public List<T> data() {

        return new ArrayList<>(mData);
    }

    public void add(final T data) {
        addAt(size(), data);
    }

    public void addAt(final int index,
                      final T data) {
        add(index, data);
    }

    public void addAll(final List<? extends T> elements) {
        mData.addAll(elements);
    }

    private void add(final int index, final T t) {
        mData.add(index, t);
    }

    public void deleteElement(final T data) {
        final int index = mData.indexOf(data);

        if (index < 0) {
            return;
        }

        deleteAt(index);
    }

    public void deleteAt(final int index) {
        mData.remove(index);
    }

    public void replaceAt(final int index, final T data) {
        mData.set(index, data);
    }

    public void move(final int from, final int to) {
        final T elt = mData.remove(from);
        mData.add(to, elt);
    }

    public void clear() {
        mData.clear();
    }

    @Override
    public int size() {
        return mData.size();
    }

    @Override
    public T get(final int index) {
        return mData.get(index);
    }

    public void insert(int index, T data) {
        mData.add(index, data);
    }
}

File: adapter/src/main/java/com/iheartradio/multitypeadapter/dataset/ListDataSet.kt
package com.iheartradio.multitypeadapter.dataset

import java.util.*

class ListDataSet<T> @JvmOverloads constructor(data: List<T> = emptyList()) : DataSet<T> {

    private val changeEvents = mutableListOf<(DataSet.ChangeEvent) -> Unit>()
    private lateinit var data: MutableList<T>

    init {
        setData(data)
    }

    private fun fireEvent(event: DataSet.ChangeEvent) {
        changeEvents.forEach { it(event) }
    }

    fun setData(newData: List<T>) {
        data = ArrayList(newData)

        fireEvent(DataSetChangeEvent())
    }

    fun data(): List<T> = data

    fun add(t: T) {
        addAt(size(), t)
    }

    fun addAt(index: Int, t: T) {
        add(index, t)
    }

    fun addAll(elements: List<T>) {
        val cntWas = data.size

        data.addAll(elements)

        val cntNow = data.size

        fireEvent(RangeAddedEvent(Range(cntWas, cntNow - 1)))
    }

    private fun add(index: Int, t: T) {
        data.add(index, t)

        fireEvent(AddedEvent(index))
    }

    fun remove(t: T) {
        val index = data.indexOf(t)

        if (index < 0) return

        removeAt(index)
    }

    fun removeAt(index: Int) {
        data.removeAt(index)

        fireEvent(RemovedEvent(index))
    }

    fun replaceAt(index: Int, t: T) {
        data[index] = t

        fireEvent(ItemChangedEvent(index))
    }

    fun move(from: Int, to: Int) {
        val itemToMove = data.removeAt(from)
        data.add(to, itemToMove)

        fireEvent(MovedEvent(from, to))
    }

    fun clear() {
        setData(emptyList())
    }

    override fun size(): Int = data.size

    override operator fun get(index: Int): T {
        return data[index]
    }

    fun addChangeEventListener(listener: (DataSet.ChangeEvent) -> Unit) {
        changeEvents.add(listener)
    }

    fun removeChangeEventListener(listener: (DataSet.ChangeEvent) -> Unit) {
        changeEvents.remove(listener)
    }

    companion object {

        fun <T> copyOf(other: DataSet<T>): ListDataSet<T> {
            return ListDataSet(DataSetUtils.listFrom(other))
        }
    }

}